### ISSUE TITLE
HLSL: mul() produce errors when the number of array elements does not match

### ISSUE THREAD

**Issue by xlpiao** [KhronosGroup/glslang#791]
```c
cbuffer ConstantBuffer : register( b0 )
{
    matrix World;
    matrix View;
    matrix Projection;
};

struct VS_INPUT
{
    float4 Pos : POSITION;
    float3 Norm : NORMAL;
};

struct PS_INPUT
{
    float4 Pos : SV_POSITION;
    float3 Norm : TEXCOORD0;
};

PS_INPUT main( VS_INPUT input )
{
    PS_INPUT output = (PS_INPUT)0;
    output.Pos = mul( input.Pos, World );
    output.Pos = mul( output.Pos, View );
    output.Pos = mul( output.Pos, Projection );
    output.Norm = mul( input.Norm, World );  // Work when output.Norm = mul( input.Norm, (float3x3)World );

    return output;
}
```

Is this gonna be processed in future, or it will be excluded?
I mean, the last element of **World** should be ignored internally on the upper case.
Is there any plan on this issue?

**Comment by steve-lunarg** [comment]
Also, there are related troubles with simple matrix truncations, such as when assigning a float4x4 to a float3x3.

It's somewhat easy to allow the matrix truncations in and of themselves, but the current function disambiguation process is not sufficient to handle the above vec*mul case with truncations, nor several other cases that have come up around e.g, vector to scalar truncations.  It needs improvement, possibly to understand that some shape conversions are better than others.

**Comment by LoopDawg** [comment]
This was partly fixed by #1161 / #1167, which left an issue with using the ConstantBuffer keyword as an identifier.  That's resolved in #1169, so the example should compile as given, after that.
